get_http.go

```diff
@@ -117,13 +118,54 @@ func (g *HttpGetter) GetFile(dst string, u *url.URL) error {
 		g.Client = httpClient
 	}
 
-	resp, err := g.Client.Get(u.String())
+	// check to see whether user has specified a range of bytes to download.
+	// if user has, but the range is invalid, fall back to downloading the
+	// whole file
+	byteRange, rangeErr := GetByteRange(u)
+	if rangeErr != nil {
+		if strings.Compare("No byte range provided", rangeErr.Error()) != 0 {
+			fmt.Printf("%s; going to download entire file without a range request",
+				rangeErr)
+		}
+	}
+
+	req, err := http.NewRequest("HEAD", u.String(), nil)
+
+	isPartialDownload := false
+	if rangeErr == nil {
+		// We first make a HEAD request so we can check if the server supports
+		// range queries. If the server/URL doesn't support HEAD requests,
+		// we just fall back to GET.
+
+		if err != nil {
+			return err
+		}
+
+		resp, err := g.Client.Do(req)
+		if err == nil && (resp.StatusCode >= 200 && resp.StatusCode < 300) {
+			// If the HEAD request succeeded, then attempt to set the range
+			// query if we can.
+			if resp.Header.Get("Accept-Ranges") == "bytes" {
+				req.Header.Set("Range", fmt.Sprintf("bytes=%s-%s",
+					byteRange[0], byteRange[1]))
+				isPartialDownload = true
+			}
+		}
+	}
+
+	// Set the request to GET now, and redo the query to download
+	req.Method = "GET"
+
+	resp, err := g.Client.Do(req)
 	if err != nil {
 		return err
 	}
 	defer resp.Body.Close()
 	if resp.StatusCode != 200 {
-		return fmt.Errorf("bad response code: %d", resp.StatusCode)
+		// if we did a ranged request we'll get a 206 instead.
+		if !((resp.StatusCode == 206) && (isPartialDownload == true)) {
```

Comment:
Complex boolean logic (not your fault, just the fact of it!). Great job commenting it to help clarify, but may I suggest making a singular if with the following?

```go
if resp.StatusCode != 200 && !(isPartialDownload && resp.StatusCode == 206)`
```

I think it reads better literally: "if not a 200 response and its not a partial download with a 206". You can further demorgan's the not into the parens but as you did above I think its easier to read in english as the above...

I **think** thats equivalent logic but like I said, tricky logic so I could be wrong!